src/pipx/venv.py

```diff
@@ -142,16 +162,46 @@ def upgrade_packaging_libraries(self, pip_args: List[str]) -> None:
         else:
             # TODO: setuptools and wheel? Original code didn't bother
             # but shared libs code does.
-            self.upgrade_package("pip", pip_args)
+            self._upgrade_package_no_metadata("pip", pip_args)
+
+    def install_package(
+        self,
+        package: str,  # if "", will be determined
+        package_or_url: str,
+        pip_args: List[str],
+        include_dependencies: bool,
+        include_apps: bool,
+        is_main_package: bool,
+    ) -> None:
+        if not package:
+            # If no package name is supplied, find out package name installed
+            #   by comparing old_package_set with new one after install
+            old_package_set = self.list_installed_packages()
```

Comment:
This is confusing. I feel like I have already mentioned this and you explained it, so my apologies, but why are we going to these lengths and adding a bunch of custom logic to get the package name? And is it 100% reliable that it gets the right package name?

Reply:
`pipx run` is the only command where we don't know for sure what the package name is.  There's no separate `package` argument in `pipx run`, just a `package_or_url`.  Even its venv directory is some hashed string, not the package name.

At first I thought I could just use any string for `package`, like `"pipx_run_package"`.  But it turns out that `get_venv_metadata()` needs the valid package name to retrieve any metadata.  And if it doesn't work, then it cannot find the `package_version`, and thus the install is labeled a failure.


As for the method I use being 100% reliable, I can't make that claim, but I cannot think of a way it would fail.  Basically we use pip to install one package and its dependencies, note all packages just installed, and find the only one of those that has no other package depend on it.  It should work, but I can't think of a way to prove it decisively.

So the options:
* Figure out the package name
* Do not verify that the package was installed in the case of `pipx run`, and thus bypass figuring out the package name.  This is what happened before, just because we never checked if the package was installed correctly in the `pipx run` code.  I moved the install success check into the venv.install_package() code, so now it always checks success whether from `pipx install`, `pipx inject`, or `pipx run`.
  * This means also we have no metadata for the `pipx run` package.

I think it's nice that the code verifies the install went properly for `pip run`, but we could do without it if we wanted.  And then we also give up valid metadata for `pipx run`.

Reply:
The other reason that we specifically discussed was that in an error condition, there's no way to make the helpful error message (See #253) if we have no metadata for the package.

Reply:
But I just had an epiphany.  We **do** know the app that the user is trying to run.  So we could just look through the installed package+dependencies and look for which one has the app inside of it.  Or which one had the app installed to the user's bin directory.  That should in the limit get closer to 100% certainty I think.  It's at least an easier metric.

Reply:
And after doing some research, there is no restriction against python packages having circular dependencies.  So if package A depends on package B, and package B depends on package A, theoretically you could end up with a situation where every package is required by every other package.

It wouldn't be a good architecture but it is possible.  So I guess that's the dependency search failure case.

Reply:
My proposal is to just use the dependency check, assume it will work 99% of the time (the other <1% being a top-level package that is part of a circular dependency). The failure mode is easy to check for--no package is found to be "unrequired".  In that case, we fall back on our current state which is no package metadata for the case of `pipx run`.

Reply:
I just ran a test of `top_of_deptree` with all of the current pipx testing packages.  "cloudtoken" has circular references, because the cloudtoken plugins require cloudtoken, and the cloudtoken requires the plugins.

So in this case, we could detect the fact that we come up empty looking for a non-required package.  It would be possible to gracefully recover, and just have no metadata like we do currently.

But 1 package out of 6 failing is a bit worse odds than I was anticipating.

Reply:
OK, one more epiphany which I need to check out. 🤞

`pip install` has a `--no-dependencies` switch.  Perhaps we can just first install the main package, see what package got installed, and THEN install everything including the dependencies.

Reply:
I think this works great.  I've committed this version.  No more top_of_deptree.